### PR TITLE
Add “searchable” capabilities to OPDS catalogs

### PR DESCRIPTION
**Pull request by bateast**
OPDS catalogs may have search query capabilities. It is very usefull to find a
book in huge catalog, even when they are organized as directories. This commit
provides the possibility to add a %s mark in an OPS catalog URL, that would
trigger a search box when browsing the OPDS in koreader.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/5823)

### CODE REVIEW

File: frontend/ui/widget/opdsbrowser.lua [hunk 1]
```diff
@@ -645,7 +660,55 @@ function OPDSBrowser:showDownloads(item)
     UIManager:show(self.download_dialog)
 end
 
+function OPDSBrowser:browse(browse_url, username, password)
+  logger.info("launch url", browse_url)
```
Comment (Frenzie) [inline frontend/ui/widget/opdsbrowser.lua]: Seems more of a logger.dbg?

File: frontend/ui/widget/opdsbrowser.lua [hunk 2]
```diff
@@ -645,7 +660,55 @@ function OPDSBrowser:showDownloads(item)
...
+  table.insert(self.paths, {
+                 url = browse_url,
+                 username = username,
+                 password = password,
+  })
+  if not self:updateCatalog(browse_url, username, password) then
+    table.remove(self.paths)
```
Comment (Frenzie) [inline frontend/ui/widget/opdsbrowser.lua]: Could you please observe the style guide? Four spaces indentation. :-) (That applies to everything in this function; I'm just attaching it to this line more or less at random.)

https://github.com/koreader/koreader-base/wiki/Coding-style

File: frontend/ui/widget/opdsbrowser.lua [hunk 3]
```diff
@@ -645,7 +660,55 @@ function OPDSBrowser:showDownloads(item)
...
+                    callback = function()
+                        UIManager:close(self.search_server_dialog)
+                    end,
+                },
+                {
+                    text = _("Search"),
+                    is_enter_default = true,
+                    callback = function()
+                                UIManager:close(self.search_server_dialog)
+                                local search, _ = self.search_server_dialog:getInputText():gsub(" ", "+")
```
Comment (Frenzie) [inline frontend/ui/widget/opdsbrowser.lua]: If you want to discard the second one you can just use `local search = function that returns multiple things`.

File: frontend/ui/widget/opdsbrowser.lua [hunk 4]
```diff
@@ -645,7 +660,55 @@ function OPDSBrowser:showDownloads(item)
...
+                                local searched_url, _ = browse_url:gsub("%%s", search)
```
Comment (Frenzie) [inline frontend/ui/widget/opdsbrowser.lua]: Same.

File: frontend/ui/widget/opdsbrowser.lua [hunk 5]
```diff
@@ -645,7 +660,55 @@ function OPDSBrowser:showDownloads(item)
...
+                                logger.info ("Searched url becomes", searched_url)
+                                self:browse(searched_url, username, password)
+                    end,
+                },
+            }
+        },
+    }
+    logger.info ("display broswse search for url", browse_url)
```
Comment (Frenzie) [inline frontend/ui/widget/opdsbrowser.lua]: Doesn't really seem necessary anymore.
```suggestion
```

File: frontend/ui/widget/opdsbrowser.lua [hunk 6]
```diff
@@ -645,7 +660,55 @@ function OPDSBrowser:showDownloads(item)
...
+    UIManager:show(self.search_server_dialog)
+    self.search_server_dialog:onShowKeyboard()
+end
+
 function OPDSBrowser:onMenuSelect(item)
+    logger.info("selected catalog item", item)
```
Comment (Frenzie) [inline frontend/ui/widget/opdsbrowser.lua]: Same as above, seems more of a more of a logger.dbg to me.

**Comment by Frenzie** [comment]
Thanks!

File: frontend/ui/widget/opdsbrowser.lua [hunk 7]
```diff
@@ -645,6 +660,51 @@ function OPDSBrowser:showDownloads(item)
...
+        table.remove(self.paths)
+    end
+end
+
+function OPDSBrowser:browseSearchable(browse_url, username, password)
+    self.search_server_dialog = InputDialog:new{
+        title = _("Search OPDS catalog"),
+        input = "",
+        hint = _("Search string"),
+        input_hint = _("author:dumas alexandre"),
```
Comment (Frenzie) [inline frontend/ui/widget/opdsbrowser.lua]: I overlooked the fact that this was put in as a translated string, but this `author:` prefix sounds like a technical thing that can't be translated?

**Comment by ptrm** [comment]
Maybe worth mentioning that the list from init()!gets overriden, not merged, by the one from `settings.reader.lua::["opds_servers”]`. That probably desirable, but might confuse users. Confused me at least, until I grepped the other one and removed it ;)

**Comment by ptrm** [comment]
But most of all, I longed for this feature for a long time :D

**Comment by ptrm** [comment]
Also, the redirects might not be supported. Or are supported, but the second download is not allowed to overwrite?

As for the overwrite popup, the result when the epub file had not existed also rendered the same results.

**Comment by Frenzie** [comment]
@ptrm It'd be better to open a new issue.

**Pull request merged**